File: writeoff-engine/cpp-reference/WriteoffEngine.cpp
#include "inventory/WriteoffEngine.hpp"
#include "core/data/DatabaseManager.hpp"
#include "core/utils/RAIIWrappers.hpp"
#include "core/utils/AppSettings.hpp"
#include "core/ActionLogger.hpp"

#include <QSettings>
#include <QUuid>
#include <QJsonObject>
#include <sqlite3.h>
#include <algorithm>
#include <numeric>
#include <string>
#include <vector>

namespace smartstock::inventory {

using core::data::DatabaseManager;
using core::data::Transaction;

namespace {
constexpr double kEps = 1e-6;

std::shared_ptr<sqlite3> connection() {
    return DatabaseManager::getInstance().getConnection();
}
} // namespace

WriteoffEngine::WriteoffEngine(QObject* parent) : QObject(parent) {}

// ─────────────────────────────────────────────────────────────────────────────
// Подбор партий (без записи в БД)
// ─────────────────────────────────────────────────────────────────────────────
WriteoffResult WriteoffEngine::computePlan(sqlite3* db, const QString& productId, double quantity,
                                           WriteoffMethod method, int warehouseId) {
    const std::string productIdStd = productId.toStdString();
    WriteoffResult r;

    struct Row { qint64 id; double avail; double cost; };
    std::vector<Row> rows;
    double totalAvail = 0.0;

    std::string sql =
        "SELECT id, (quantity - reserved) AS avail, cost_price FROM batches "
        "WHERE product_id = ?1 AND status = 'active' "
        "AND (quantity - reserved) > 1e-6 "
        "AND (?2 = 0 OR warehouse_id = ?2)";
    switch (method) {
        case WriteoffMethod::FIFO: sql += " ORDER BY received_at ASC, id ASC;";  break;
        case WriteoffMethod::LIFO: sql += " ORDER BY received_at DESC, id DESC;"; break;
        case WriteoffMethod::WeightedAverage: sql += ";"; break;
    }

    try {
        utils::SQLiteStatement stmt(db, sql.c_str());
        sqlite3_bind_text(stmt, 1, productIdStd.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_bind_int(stmt, 2, warehouseId);
        while (sqlite3_step(stmt) == SQLITE_ROW) {
            Row row;
            row.id    = sqlite3_column_int64(stmt, 0);
            row.avail = sqlite3_column_double(stmt, 1);
            row.cost  = sqlite3_column_double(stmt, 2);
            rows.push_back(row);
            totalAvail += row.avail;
        }
    } catch (const std::exception& e) {
        r.success   = false;
        r.errorCode = "DB_ERROR";
        r.errorMsg  = QString::fromStdString(e.what());
        return r;
    }

    // Недостаточно остатка.
    if (totalAvail + kEps < quantity) {
        r.success   = false;
        r.errorCode = "INSUFFICIENT_STOCK";
        r.errorMsg  = QStringLiteral("Недостаточно остатка: доступно %1, запрошено %2")
                          .arg(totalAvail).arg(quantity);
        return r;
    }

    if (method == WriteoffMethod::WeightedAverage) {
        // Средневзвешенная цена доступного остатка.
        const double weighted = std::accumulate(rows.begin(), rows.end(), 0.0,
            [](double s, const Row& r) { return s + r.cost * r.avail; });
        const double avg = (totalAvail > kEps) ? (weighted / totalAvail) : 0.0;

        // Списываем из всех партий пропорционально их доле; остаток округления
        // отдаём последней значимой строке, чтобы сумма совпала с quantity.
        double allocated = 0.0;
        std::size_t lastIdx = rows.size();
        for (std::size_t i = 0; i < rows.size(); ++i) {
            if (rows[i].avail > kEps) lastIdx = i;
        }
        for (std::size_t i = 0; i < rows.size(); ++i) {
            if (rows[i].avail <= kEps) continue;
            double share = (i == lastIdx)
                               ? (quantity - allocated)
                               : (quantity * rows[i].avail / totalAvail);
            if (share <= kEps) continue;
            allocated += share;
            WriteoffLine line;
            line.batchId    = rows[i].id;
            line.quantity   = share;
            line.costPrice  = avg;
            line.lineAmount = share * avg;
            r.lines.push_back(line);
            r.totalCost += line.lineAmount;
        }
        r.avgCost = avg;
    } else {
        // FIFO / LIFO: берём партии по порядку, пока не наберём quantity.
        double remaining = quantity;
        for (const auto& row : rows) {
            if (remaining <= kEps) break;
            double take = std::min(remaining, row.avail);
            if (take <= kEps) continue;
            WriteoffLine line;
            line.batchId    = row.id;
            line.quantity   = take;
            line.costPrice  = row.cost;
            line.lineAmount = take * row.cost;
            r.lines.push_back(line);
            r.totalCost += line.lineAmount;
            remaining   -= take;
        }
        r.avgCost = (quantity > kEps) ? (r.totalCost / quantity) : 0.0;
    }

    r.success = true;
    return r;
}

// ─────────────────────────────────────────────────────────────────────────────
// Списание с записью в БД
// ─────────────────────────────────────────────────────────────────────────────
WriteoffResult WriteoffEngine::writeoffInTransaction(sqlite3* db, const QString& productId,
                                                     double quantity, WriteoffMethod method,
                                                     const QString& reason, qint64 documentId,
                                                     int warehouseId) {
    WriteoffResult r = computePlan(db, productId, quantity, method, warehouseId);
    if (!r.success) {
        return r;  // INSUFFICIENT_STOCK/DB_ERROR — транзакция вызывающего откатит
    }

    const std::string reasonStd = reason.toStdString();

    for (const WriteoffLine& line : r.lines) {
        // Уменьшаем остаток партии; при обнулении помечаем 'exhausted'.
        utils::SQLiteStatement upd(db,
            "UPDATE batches "
            "SET quantity = quantity - ?1, "
            "    status = CASE WHEN (quantity - ?1) <= 1e-6 THEN 'exhausted' ELSE status END "
            "WHERE id = ?2;");
        sqlite3_bind_double(upd, 1, line.quantity);
        sqlite3_bind_int64 (upd, 2, line.batchId);
        if (sqlite3_step(upd) != SQLITE_DONE) {
            throw std::runtime_error(std::string("UPDATE batches: ") + sqlite3_errmsg(db));
        }

        // Фиксируем движение списания.
        utils::SQLiteStatement ins(db,
            "INSERT INTO batch_writeoffs"
            "(id, batch_id, quantity, cost_price, reason, document_id, written_by) "
            "VALUES (?1, ?2, ?3, ?4, ?5, ?6, NULL);");
        const std::string woId =
            QUuid::createUuid().toString(QUuid::WithoutBraces).toStdString();
        sqlite3_bind_text  (ins, 1, woId.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_bind_int64 (ins, 2, line.batchId);
        sqlite3_bind_double(ins, 3, line.quantity);
        sqlite3_bind_double(ins, 4, line.costPrice);
        sqlite3_bind_text  (ins, 5, reasonStd.c_str(), -1, SQLITE_TRANSIENT);
        if (documentId > 0) sqlite3_bind_int64(ins, 6, documentId);
        else                sqlite3_bind_null (ins, 6);
        if (sqlite3_step(ins) != SQLITE_DONE) {
            throw std::runtime_error(std::string("INSERT batch_writeoffs: ") + sqlite3_errmsg(db));
        }
    }

    if (documentId > 0) {
        utils::SQLiteStatement doc(db,
            "UPDATE documents SET updated_at = strftime('%s','now') WHERE id = ?1;");
        sqlite3_bind_int64(doc, 1, documentId);
        if (sqlite3_step(doc) != SQLITE_DONE) {
            throw std::runtime_error(std::string("UPDATE documents: ") + sqlite3_errmsg(db));
        }
    }

    r.success = true;
    return r;
}

void WriteoffEngine::postWriteoffEffects(const QString& productId, double quantity,
                                         const QString& reason, qint64 documentId,
                                         int warehouseId, const WriteoffResult& r) {
    QJsonObject v;
    v["product"]  = productId;
    v["quantity"] = quantity;
    v["total"]    = r.totalCost;
    v["reason"]   = reason;
    smartstock::core::ActionLogger::instance().logCreate("batch_writeoff", documentId, v);

    // Уведомления об остатке (после фиксации).
    const double remaining = availableQuantity(productId, warehouseId);
    const int threshold = smartstock::core::utils::lowStockThreshold();
    if (remaining < threshold) emit lowStockWarning(productId, remaining);
    if (remaining < kEps)      emit stockDepleted(productId);
}

WriteoffResult WriteoffEngine::writeoff(const QString& productId, double quantity, WriteoffMethod method,
                                        const QString& reason, qint64 documentId, int warehouseId) {
    WriteoffResult r;
    auto dbSp = connection();
    sqlite3* db = dbSp.get();
    if (!db) {
        r.errorCode = "DB_ERROR";
        r.errorMsg  = QStringLiteral("Нет соединения с БД");
        return r;
    }

    try {
        Transaction tx(db);
        r = writeoffInTransaction(db, productId, quantity, method, reason, documentId, warehouseId);
        if (!r.success) {
            return r;  // INSUFFICIENT_STOCK/DB_ERROR — Transaction откатит изменения
        }
        tx.commit();
    } catch (const std::exception& e) {
        WriteoffResult err;
        err.success   = false;
        err.errorCode = "DB_ERROR";
        err.errorMsg  = QString::fromStdString(e.what());
        return err;
    }

    postWriteoffEffects(productId, quantity, reason, documentId, warehouseId, r);
    return r;
}

// ─────────────────────────────────────────────────────────────────────────────
// Симуляция (только чтение)
// ─────────────────────────────────────────────────────────────────────────────
WriteoffResult WriteoffEngine::simulate(const QString& productId, double quantity,
                                        WriteoffMethod method, int warehouseId) {
    auto dbSp = connection();
    sqlite3* db = dbSp.get();
    if (!db) {
        WriteoffResult r;
        r.errorCode = "DB_ERROR";
        r.errorMsg  = QStringLiteral("Нет соединения с БД");
        return r;
    }
    return computePlan(db, productId, quantity, method, warehouseId);
}

// ─────────────────────────────────────────────────────────────────────────────
// Справочные расчёты
// ─────────────────────────────────────────────────────────────────────────────
double WriteoffEngine::weightedAverageCost(const QString& productId, int warehouseId) {
    const std::string productIdStd = productId.toStdString();
    auto dbSp = connection();
    sqlite3* db = dbSp.get();
    if (!db) return 0.0;
    try {
        utils::SQLiteStatement stmt(db,
            "SELECT CASE WHEN SUM(quantity - reserved) > 1e-6 "
            "            THEN SUM(cost_price * (quantity - reserved)) / SUM(quantity - reserved) "
            "            ELSE 0 END "
            "FROM batches "
            "WHERE product_id = ?1 AND status = 'active' "
            "AND (quantity - reserved) > 1e-6 "
            "AND (?2 = 0 OR warehouse_id = ?2);");
        sqlite3_bind_text(stmt, 1, productIdStd.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_bind_int(stmt, 2, warehouseId);
        if (sqlite3_step(stmt) == SQLITE_ROW) {
            return sqlite3_column_double(stmt, 0);
        }
    } catch (const std::exception&) {
        // невалидное соединение/SQL — возвращаем 0
    }
    return 0.0;
}

double WriteoffEngine::availableQuantity(const QString& productId, int warehouseId) {
    const std::string productIdStd = productId.toStdString();
    auto dbSp = connection();
    sqlite3* db = dbSp.get();
    if (!db) return 0.0;
    try {
        utils::SQLiteStatement stmt(db,
            "SELECT COALESCE(SUM(quantity - reserved), 0) FROM batches "
            "WHERE product_id = ?1 AND status = 'active' "
            "AND (?2 = 0 OR warehouse_id = ?2);");
        sqlite3_bind_text(stmt, 1, productIdStd.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_bind_int(stmt, 2, warehouseId);
        if (sqlite3_step(stmt) == SQLITE_ROW) {
            return sqlite3_column_double(stmt, 0);
        }
    } catch (const std::exception&) {
        // невалидное соединение/SQL — возвращаем 0
    }
    return 0.0;
}

} // namespace smartstock::inventory

File: writeoff-engine/src/writeoff.engine.ts
// Движок списания партий (порт WriteoffEngine из десктопа, src/inventory/WriteoffEngine.cpp).
//
// Storage-agnostic: работает не с конкретной БД, а через интерфейс BatchStore.
// Семантика 1:1 с C++ оригиналом:
//  - подбор партий: status='active' И (quantity - reserved) > eps, фильтр по складу;
//  - FIFO: ORDER BY received_at ASC, id ASC; LIFO: received_at DESC, id DESC;
//  - WeightedAverage: НЕ списывает физически FIFO — распределяет qty ПРОПОРЦИОНАЛЬНО
//    долям доступного остатка всех активных партий, cost_price каждой строки =
//    средневзвешенная цена; остаток округления отдаётся последней значимой партии
//    (как computePlan() в C++);
//  - недостаточно остатка → INSUFFICIENT_STOCK, БЕЗ частичного списания (вся
//    транзакция откатывается);
//  - обнулённая партия → status='exhausted';
//  - каждое списание → запись batch_writeoffs;
//  - после — декремент кэша products.quantity.
//
// Транзакционность обеспечивает вызывающий код: applyWriteoff выполняется уже
// ВНУТРИ открытой транзакции переданного BatchStore. При INSUFFICIENT_STOCK
// бросается исключение, чтобы внешняя транзакция откатилась без частичного списания.

import {
  EPS,
  type BatchStore,
  type WriteoffMethod,
  type WriteoffReason,
  type WriteoffResult,
  type WriteoffLineResult,
  type BatchRow,
} from './types';

// Округление денег до 2 знаков (как toFixed(2) семантика для NUMERIC(15,2)).
function round2(v: number): number {
  return Math.round((v + Number.EPSILON) * 100) / 100;
}

// Доступный остаток партии (как (quantity - reserved) в SQL C++).
function available(b: BatchRow): number {
  return b.quantity - b.reserved;
}

export interface WriteoffParams {
  productId: string;
  warehouseId?: string;
  quantity: number;
  reason: WriteoffReason;
  method: WriteoffMethod;
  documentId?: string | null;
  userId: string;
}

// ─────────────────────────────────────────────────────────────────────────────
// Подбор партий + расчёт строк списания (как WriteoffEngine::computePlan).
// Только расчёт, без записи. success=false при нехватке остатка.
// ─────────────────────────────────────────────────────────────────────────────
export async function computePlan(
  store: BatchStore,
  productId: string,
  quantity: number,
  method: WriteoffMethod,
  warehouseId?: string,
): Promise<WriteoffResult> {
  const order =
    method === 'FIFO'
      ? 'received_asc'
      : method === 'LIFO'
        ? 'received_desc'
        : 'none';

  const rows = await store.findActiveBatches({ productId, warehouseId }, order);

  const totalAvail = rows.reduce((s, r) => s + available(r), 0);

  // Недостаточно остатка (как totalAvail + kEps < quantity).
  if (totalAvail + EPS < quantity) {
    return {
      success: false,
      totalCost: 0,
      averageCost: 0,
      items: [],
      error: 'INSUFFICIENT_STOCK',
      errorMessage: `Недостаточно остатка: доступно ${totalAvail}, запрошено ${quantity}`,
    };
  }

  const items: WriteoffLineResult[] = [];
  let totalCost = 0;

  if (method === 'WeightedAverage') {
    // Средневзвешенная цена доступного остатка (как weighted/totalAvail в C++).
    const weighted = rows.reduce((s, r) => s + r.costPrice * available(r), 0);
    const avg = totalAvail > EPS ? weighted / totalAvail : 0;

    // Индекс последней значимой партии — ей отдаём остаток округления.
    let lastIdx = rows.length;
    for (let i = 0; i < rows.length; i++) {
      if (available(rows[i]!) > EPS) lastIdx = i;
    }

    let allocated = 0;
    for (let i = 0; i < rows.length; i++) {
      const row = rows[i]!;
      const avail = available(row);
      if (avail <= EPS) continue;
      let share =
        i === lastIdx ? quantity - allocated : (quantity * avail) / totalAvail;
      // Не списываем из партии больше, чем в ней есть (иначе остаток партии ушёл
      // бы в минус из-за добора округления в последнюю строку).
      share = Math.min(share, avail);
      if (share <= EPS) continue;
      allocated += share;
      const cost = round2(avg);
      items.push({
        batchId: row.id,
        quantity: share,
        costPrice: cost,
        costCurrency: row.costCurrency,
        fxRate: row.fxRate,
      });
      totalCost += share * cost;
    }
    return {
      success: true,
      items,
      totalCost: round2(totalCost),
      averageCost: round2(avg),
    };
  }

  // FIFO / LIFO: берём партии по порядку, пока не наберём quantity.
  let remaining = quantity;
  for (const row of rows) {
    if (remaining <= EPS) break;
    const take = Math.min(remaining, available(row));
    if (take <= EPS) continue;
    const cost = round2(row.costPrice);
    items.push({
      batchId: row.id,
      quantity: take,
      costPrice: cost,
      costCurrency: row.costCurrency,
      fxRate: row.fxRate,
    });
    totalCost += take * cost;
    remaining -= take;
  }
  const avgCost = quantity > EPS ? totalCost / quantity : 0;
  return {
    success: true,
    items,
    totalCost: round2(totalCost),
    averageCost: round2(avgCost),
  };
}

// ─────────────────────────────────────────────────────────────────────────────
// Расчёт + применение к хранилищу через переданный store.
// Вызывается уже ВНУТРИ транзакции. При INSUFFICIENT_STOCK бросает, чтобы
// внешняя транзакция откатилась без частичного списания (как Transaction в C++).
// ─────────────────────────────────────────────────────────────────────────────
export async function applyWriteoff(
  store: BatchStore,
  params: WriteoffParams,
): Promise<WriteoffResult> {
  const { productId, warehouseId, quantity, method, reason, documentId, userId } =
    params;

  const plan = await computePlan(store, productId, quantity, method, warehouseId);
  if (!plan.success) {
    // Бросаем, чтобы откатить транзакцию (никаких частичных изменений).
    const err = new Error(plan.errorMessage ?? plan.error ?? 'writeoff failed');
    (err as Error & { code?: string }).code = plan.error;
    throw err;
  }

  for (const line of plan.items) {
    // Уменьшаем остаток партии; при обнулении → status='exhausted'.
    await store.decrementBatchQuantity(line.batchId, line.quantity);
    // Фиксируем движение списания.
    await store.insertWriteoff({
      batchId: line.batchId,
      quantity: line.quantity,
      costPrice: line.costPrice,
      reason,
      documentId: documentId ?? null,
      userId,
    });
  }

  // Обновляем денормализованный кэш остатка товара.
  await store.decrementProductQuantity(productId, quantity);

  return plan;
}

// ─────────────────────────────────────────────────────────────────────────────
// Справочные расчёты (как availableQuantity / weightedAverageCost в C++).
// Только чтение — работают через store.
// ─────────────────────────────────────────────────────────────────────────────

export async function getAvailableQtyVia(
  store: BatchStore,
  productId: string,
  warehouseId?: string,
): Promise<number> {
  // Все активные партии (status='active'), сумма (quantity - reserved).
  const rows = await store.findActiveBatches({ productId, warehouseId }, 'none');
  return rows.reduce((s, r) => s + available(r), 0);
}

export async function getWeightedAverageCostVia(
  store: BatchStore,
  productId: string,
  warehouseId?: string,
): Promise<number> {
  const rows = await store.findActiveBatches({ productId, warehouseId }, 'none');
  const totalAvail = rows.reduce((s, r) => s + available(r), 0);
  if (totalAvail <= EPS) return 0;
  const weighted = rows.reduce((s, r) => s + r.costPrice * available(r), 0);
  return round2(weighted / totalAvail);
}

export { round2 };
